### ISSUE TITLE
OP/ED misidentification issue

### ISSUE THREAD

**Issue by purposelycryptic** [ZeroQI/Absolute-Series-Scanner#172]
First of all, thank you for all your hard work on HAMA/ASS.

This is a rather odd issue I've only recently noticed, after finally figuring out how to get AniAdd to properly rename OP/EDs with ASS-compatible numbering: 

For some reason, ASS will misidentify certain OP/ED files, usually identifying them as the OP/ED above or below the correct one in AniDB's internal ordering (i.e. episode C3 will be IDed as episode C2 or C4 instead).

My Files are all organized/named according to the following scheme:
 - Series_Name [Group_Name] [Source, Resolution]\
	 - Series_Name - Ep_Num [Group_Name] [Source, Resolution, Audio_Codec] [CRC_32].Ext
	 
Where _Ep_Num_ is the AniDB Episode Number (Specials using the _SP_ prefix, OPs and EDs the _OP_ and _ED_ prefix, respectively, and an _[a-z]_ suffix when necessary, Trailer/Previews the _T_ prefix, and non-systematic Other files the _O_ prefix), with multi-part episodes having a ' _- partX_' suffix, and conjoined episodes having a '_-XX_' suffix, where _X_/_XX_ is the respective part/episode number.

As an example, here is the file/folder layout for _Owarimonogatari (2017)_:
 - Owarimonogatari (2017) [Dmon] [BD, 720p]\
	 - Owarimonogatari (2017) - 01 - part1 [Dmon] [BD, 720p, AAC] [A648D7B1].mkv
	 - Owarimonogatari (2017) - 01 - part2 [Dmon] [BD, 720p, AAC] [534EAEF0].mkv
	 - Owarimonogatari (2017) - 02 [Dmon] [BD, 720p, AAC] [8A34B664].mkv
	 - Owarimonogatari (2017) - ED1c [Dmon] [BD, 720p, AAC] [702EA02D].mkv
	 - Owarimonogatari (2017) - ED1d [Dmon] [BD, 720p, AAC] [930DFFFA].mkv
	 - Owarimonogatari (2017) - ED1e [Dmon] [BD, 720p, AAC] [66EC229F].mkv
	 - Owarimonogatari (2017) - OP1b [Dmon] [BD, 720p, AAC] [C7E110E0].mkv
	 - Owarimonogatari (2017) - OP2 [Dmon] [BD, 720p, AAC] [63B59143].mkv
	 - Owarimonogatari (2017) - OP3 [Dmon] [BD, 720p, AAC] [C0D815AE].mkv

And this is what the _Specials_ look like in Plex:
![OwariSpecials1](https://i.imgur.com/z06Bx4B.jpg)

Here the EDs were correctly IDed, but as for the OPs, they were instead matched like this:
![OwariSpecials2](https://i.imgur.com/vuWrulN.jpg)

![OwariSpecials3](https://i.imgur.com/jPwaSfQ.jpg)

I'm not entirely sure what factor could be causing the issue, as it seems to be somewhat random - in some series, all OP/EDs are correctly matched, in some only OPs or EDs, and in some a few of both. There is obviously some common factor at work, but it eludes me. 

Up until recently, all my OP/EDs were numbered according to AniDB's internal scheme (C1, C2, etc), simply because that is what AniAdd uses, and thus weren't identifiable by ASS and instead labeled as 5XX-numbered episodes. But after much script-wrangling, I finally managed to get it to name them with ASS-compatible numbering, or so I thought. If the issue is on my end, I apologize for taking up your time, and would appreciate any advice you could give to correct it.

Thanks!

**Comment by ZeroQI** [comment]
Please update (major scanner update done then this fix). Please report any difference

**Comment by purposelycryptic** [comment]
Using Owarimonogatari (2017) as my test case again, I first tried refreshing the metadata for the series, and unmatching/rematching it, both of which gave the same result as above (Even after unmatching, OP1b and OP2 were still shown in Plex as versions of the same episode, with no option to split).

So I created a new library containing only the series, which resulted in a different issue:

![owarispec4](https://i.imgur.com/QFuUwvR.jpg)

ASS identified OP1b, OP2 and OP3 as episodes 106, 107 and 108, respectively. As you can see below, the series only has four OPs on AniDB, total:

![Imgur](https://i.imgur.com/qw79P9Y.jpg)

I'm not sure what variable from AniDB you are using to match OP/EDs, but I'm guessing you're likely doing something similar to what my AniAdd renaming script does (except in reverse), that is, cycling through an array of the XMLs episode elements to match the episode number prefixes _OP_ and _ED_ (and their variations) to AniDB's internal episode type 3 (internal prefix _C_), then using the _episode.title_ variable to match _OP1_/_ED1_ to "_Opening_" or "_Opening 1_"/"_Ending_" or "_Ending 1_" respectively, and _OP_/_ED_(_n_)[_a-z_] to "_Opening_ (_n_)[_a-z_]"/"_Ending_ (_n_)[_a-z_]" for the rest?

At least, that is the only way I can think of, given the limited information in AniDB's series XML files, to match them based on OP/ED number, since the internal categorization/numbering scheme doesn't distinguish between OPs and EDs at all (Off-topic, but it would be awesome if ASS could eventually identify OP/ED and "Other" files by AniDB's internal numbering, i.e. C1, C2, O1, O2, etc as well).

I don't really know what scheme you use to assign them an episode number in Plex, though (aside from the category offsets), that is, which episode becomes 101, 102, etc, since AniDB's C category (type 3) numbering frequently does not reflect the OP/ED numbering order - for _Owarimonogatari (2017)_, for example, OP1b, OP2 and OP3 are C3, C2 and C4, respectively, and, in the latest version of ASS, were assigned to 106, 107 and 108.

If you are assigning the episode numbers by alphabetical order of the _episode.title_ variable (in this case, OP1a, OP1b, OP2, OP3), which seems to be case, could the current situation simply be an offset issue? I.e. instead of OP1a -> 101, OP1b -> 102, etc, it is assigning them stating at 105, hence OP1b being 106, OP2 107 and OP3 108, but still associating the metadata with the correct numbering? It seems like an odd offset miss (4), but it would match the results.

Anyway, I suppose that is enough guesswork for now, hope it is in some way helpful, and, as always, thanks for all the hard work :-) I may try and take a look at the code myself, but my Python experience is essentially limited to looking at HAMA and the Shoko Metadata Agent, so I'm very much a beginner.

**Comment by ZeroQI** [comment]
Lines 804-814 responsible for that.

From scanner logs
```
No forced guid found in folder name nor id file
misc_count: {'op2': 1, 'op3': 1, 'ed1d': 1, 'ed1e': 1, 'op1b': 1, 'ed1c': 1}
"Owarimonogatari (2017)" s00e153                         "ANIDB_RX-2" "" "Owarimonogatari (2017) - ED1c [Dmon] [BD, 720p, AAC] [702EA02D].mkv"
"Owarimonogatari (2017)" s00e154                         "ANIDB_RX-2" "" "Owarimonogatari (2017) - ED1d [Dmon] [BD, 720p, AAC] [930DFFFA].mkv"
"Owarimonogatari (2017)" s00e155                         "ANIDB_RX-2" "" "Owarimonogatari (2017) - ED1e [Dmon] [BD, 720p, AAC] [66EC229F].mkv"
"Owarimonogatari (2017)" s00e106                         "ANIDB_RX-1" "" "Owarimonogatari (2017) - OP1b [Dmon] [BD, 720p, AAC] [C7E110E0].mkv"
"Owarimonogatari (2017)" s00e107                         "ANIDB_RX-1" "" "Owarimonogatari (2017) - OP2 [Dmon] [BD, 720p, AAC] [63B59143].mkv"
"Owarimonogatari (2017)" s00e108                         "ANIDB_RX-1" "" "Owarimonogatari (2017) - OP3 [Dmon] [BD, 720p, AAC] [C0D815AE].mkv"
```
Before they were not shifted if 1b, etc... were present before 2, 3, ...
Now ending shifts it since I sorted alphabetically but put ending first
need to sort per special type per ep group so sum(dict.values()) represents total offsets to date
This should do the trick
```
misc_count: {'op2': 1, 'op3': 1, 'ed1d': 1, 'ed1e': 1, 'op1b': 1, 'ed1c': 1}
"Owarimonogatari (2017)" s00e153                         "ANIDB_RX-2" "" "Owarimonogatari (2017) - ED1c [Dmon] [BD, 720p, AAC] [702EA02D].mkv"
"Owarimonogatari (2017)" s00e154                         "ANIDB_RX-2" "" "Owarimonogatari (2017) - ED1d [Dmon] [BD, 720p, AAC] [930DFFFA].mkv"
"Owarimonogatari (2017)" s00e155                         "ANIDB_RX-2" "" "Owarimonogatari (2017) - ED1e [Dmon] [BD, 720p, AAC] [66EC229F].mkv"
"Owarimonogatari (2017)" s00e102                         "ANIDB_RX-1" "" "Owarimonogatari (2017) - OP1b [Dmon] [BD, 720p, AAC] [C7E110E0].mkv"
"Owarimonogatari (2017)" s00e103                         "ANIDB_RX-1" "" "Owarimonogatari (2017) - OP2 [Dmon] [BD, 720p, AAC] [63B59143].mkv"
"Owarimonogatari (2017)" s00e104                         "ANIDB_RX-1" "" "Owarimonogatari (2017) - OP3 [Dmon] [BD, 720p, AAC] [C0D815AE].mkv"
```
Please update and report and if not solved paste series specific scanner logs located in Plex Media Server\Plug-in Support\Data\com.plexapp.agents.hama\DataItems\_Logs
If solved please close and consider donation if not already done

**Comment by EndOfLine369** [comment]
Crashes.
File: `[Chihiro]_Amagami_SS_-_ED1_[1280x720_Blu-Ray_FLAC][970F7DCA].mkv`
```python
Error in Python: Running scanner:
Traceback (most recent call last):
  File "/share/CACHEDEV1_DATA/.qpkg/PlexMediaServer/Library/Plex Media Server/Scanners/Series/Absolute Series Scanner.py", line 810, in Scan
    offset += sum( AniDB_op[ANIDB_RX.index(rx)].values() )
KeyError: (2,)
```

**Comment by purposelycryptic** [comment]
Not solved, and seems to have generated other issues (Creating a new library, only half the series show up in Plex, and previously correctly IDed series misidentified).

Results this time, creating new Library with same series as example as previously (And leaving aside the other series, which were included as samples for the [AniDB Series Poster Priority/Ranking Issue](https://github.com/ZeroQI/Hama.bundle/issues/245) with HAMA), are as such:

![owarispecials6](https://i.imgur.com/T9hgzMV.png)

Essentially, the episode number in Plex now matches AniDB's internal C episode number and name, but is still being associated with the wrong file.

So,
![Imgur](https://i.imgur.com/T9PTqQc.gif)

...given a big Pull Request was added while I was writing this and testing things for this and the issue with HAMA, I'm not sure if any of this is still relevant or helpful, however.

I've attached the logs for the series - some of the data may be from the previous test-version of ASS, as, while I created a new Library each time, it was always named 'test1', so the logs were written to the same folder.

I've also uploaded the logs for the entire 'test1' Library (15 series) to a Google Drive folder [here](https://drive.google.com/open?id=1uyGTmk9Qo7N-P9Ad_SQiEIRTEXbga2zP), in part for the other issue, but they might prove helpful here as well, at least in giving a hint as to what went wrong with the series identification.

[Owarimonogatari (2017) [Dmon] [BD, 720p].agent-search.log](https://github.com/ZeroQI/Absolute-Series-Scanner/files/2300792/Owarimonogatari.2017.Dmon.BD.720p.agent-search.log)
[Owarimonogatari (2017) [Dmon] [BD, 720p].agent-update.log](https://github.com/ZeroQI/Absolute-Series-Scanner/files/2300793/Owarimonogatari.2017.Dmon.BD.720p.agent-update.log)
[Owarimonogatari (2017) [Dmon] [BD, 720p].filelist.log](https://github.com/ZeroQI/Absolute-Series-Scanner/files/2300794/Owarimonogatari.2017.Dmon.BD.720p.filelist.log)
[Owarimonogatari (2017) [Dmon] [BD, 720p].scanner.log](https://github.com/ZeroQI/Absolute-Series-Scanner/files/2300795/Owarimonogatari.2017.Dmon.BD.720p.scanner.log)

**Comment by purposelycryptic** [comment]
Ok, this is post-commit [306101e58727e3b542a96876fa506b0b55615cc1](https://github.com/ZeroQI/Absolute-Series-Scanner/commit/306101e58727e3b542a96876fa506b0b55615cc1), using the new version of ASS:

Deleted/recreated Library, but same OP/ED issue remains, as well as the issue of half the series in the Library's folder not showing up in Plex, and one series being misidentified as its own sequel, which it had not been, using earlier revisions (I personally started seeing series identification issues after commit  [2ca520f1ae8241c0251e082d85911654489975ce](https://github.com/ZeroQI/Absolute-Series-Scanner/commit/2ca520f1ae8241c0251e082d85911654489975ce), though I haven't had a chance to look into it as of yet)

I've uploaded the new logs for the Library, as well as the HAMA log, to a Google Drive folder [here](https://drive.google.com/open?id=1tkirrjt1vtqWCjm_-4Dbr66XLa7ZrIW_), and am attaching the logs specific to the series I have been using for testing to this post as well.

All the best!

[Owarimonogatari (2017) [Dmon] [BD, 720p].filelist.log](https://github.com/ZeroQI/Absolute-Series-Scanner/files/2300869/Owarimonogatari.2017.Dmon.BD.720p.filelist.log)
[Owarimonogatari (2017) [Dmon] [BD, 720p].scanner.log](https://github.com/ZeroQI/Absolute-Series-Scanner/files/2300870/Owarimonogatari.2017.Dmon.BD.720p.scanner.log)

**Comment by purposelycryptic** [comment]
Just FYI, I rolled back to an older version of ASS (Commit be48f452bd654550031cc18dcf990e2f90b981aa), and all the missing series in my test library were detected - of course, the OP/EDs were back where we started, but it seems like something in the first of those two big recent merges caused the series detection issue (or something right before it, as I've updated ASS fairly regularly, and the first time it happened to me was after switching to that version).

Don't know if that is helpful information or not, but I'm a firm believer in there being no such thing as too much information. And now, to bed :-)

**Comment by ZeroQI** [comment]
@purposelycryptic That should do it. It took quite a while to spot as most times when the scanner crashes, it doesn't tell you where...

```
"Owarimonogatari (2017)" s00e153                         "ANIDB_RX-2" "Ending 1c" "Owarimonogatari (2017) - ED1c [Dmon] [BD, 720p, AAC] [702EA02D].mkv"
"Owarimonogatari (2017)" s00e154                         "ANIDB_RX-2" "Ending 1d" "Owarimonogatari (2017) - ED1d [Dmon] [BD, 720p, AAC] [930DFFFA].mkv"
"Owarimonogatari (2017)" s00e155                         "ANIDB_RX-2" "Ending 1e" "Owarimonogatari (2017) - ED1e [Dmon] [BD, 720p, AAC] [66EC229F].mkv"
"Owarimonogatari (2017)" s00e102                         "ANIDB_RX-1" "Opening 1b" "Owarimonogatari (2017) - OP1b [Dmon] [BD, 720p, AAC] [C7E110E0].mkv"
"Owarimonogatari (2017)" s00e103                         "ANIDB_RX-1" "Opening 2" "Owarimonogatari (2017) - OP2 [Dmon] [BD, 720p, AAC] [63B59143].mkv"
"Owarimonogatari (2017)" s00e104                         "ANIDB_RX-1" "Opening 3" "Owarimonogatari (2017) - OP3 [Dmon] [BD, 720p, AAC] [C0D815AE].mkv"
```
Looks better now and no longer crashes. Please test and report

**Comment by ZeroQI** [comment]
I fixed the obvious, the specials numbering in ASS, which now look grand' BUT relies on files present so if missing 1a, 1b, 1c, OP2 is 102... need to download anidb xml from ASS which i wanted to avoid...
Need to match logic on hama side now

**Comment by purposelycryptic** [comment]
I'm actually seriously impressed ASS has done such a good job so far in almost every case _without_ referencing a series' AniDB XML - it's essentially like working blind with a 99% success rate. 

But, given a the majority of releases don't contain a full set of OP/EDs (or, rather, the OP/ED list on AniDB tries to cover every variation from every format the series was released in, so different media-type groups simply don't have them to release), using the XML is probably the only way to achieve full accuracy. I don't suppose it's as simple as using the same AniDB XML storage folder for both ASS and HAMA, so that, for either one, if a sufficiently recent version is present locally, it can use that vs. downloading a fresh copy... Or rather, even if it is, the larger part is then integrating the use of the XML data into ASS.

Knowing that definitely explains a lot about the code and how it is structured, though - I spent a couple hours yesterday reading over it (HAMA/ASS have been my first major exposure to Python outside of minor script editing), and somehow completely missed the fact that the XML is never fetched/referenced.

Up until a few months ago, I had been using Emby/Mediabrowser for almost a decade, and Emby Server uses it's own scanner, with plugins only able to serve as agents, so I never got a real look at the scanner side; for what it's worth, HAMA was essentially the reason I switched to Plex, despite preferring Emby over Plex in almost every other respect.

**Comment by ZeroQI** [comment]
AniDB XML example: http://api.anidb.net:9001/httpapi?request=anime&client=hama&clientver=1&protover=1&aid=13033
- epno type="3" = C1, title = Opening 1a
- epno type="3" = C2, title = Opening 2
- epno type="3" = C3, title = Opening 1b
- epno type="3" = C4, title = Opening 3
- epno type="3" = C5, title = Ending 1a
- epno type="3" = C6, title = Ending 1b
- epno type="3" = C7, title = Ending 1c
- epno type="3" = C8, title = Ending 1d
- epno type="3" = C9, title = Ending 1e
- epno type="3" = C10, title = Ending 1f

Solutions:
- leave as is as works for most files (Hama still needs updating)
- ask to rename files from OP1b to s00e102 for example
- anidb id in series name + anidb xml downloaded from here (otherwise new files mapped wrong on first scan as not downloaded if using hama cache)
- numbering with letter not sequential (see http://anidb.net/perl-bin/animedb.pl?show=anime&aid=13033 listiong op1a op2 op1b)

Still researching how to solve efficiently

**Comment by ZeroQI** [comment]
@purposelycryptic re-did the op/ed part. 

Issue was, if op1a and op1b not present, op2 is ep 2 (+100/150)...
If the numbering is off compare to what it should, force the anidb id.

Please test and report, but this is for the scanner part only. 
Once all is confirmed ok will change the agent part if needed.

https://github.com/ZeroQI/Absolute-Series-Scanner/blob/master/Scanners/Series/Absolute%20Series%20Scanner.py

New house with mortgage starting any day so time is scarce.
I am glad the agent is good enough for you, it is way more complex than other anime agents but tried to make it the best and glad it is used.

**Comment by EndOfLine369** [comment]
@ZeroQI, doesn't this new dl of anidb info `urlopen(ANIDB_HTTP_API_URL+id, context=SSL_CONTEXT).read()` look like a prime candidate of an easy AniDB ban?

**Comment by EndOfLine369** [comment]
EX: I have 110 series with op/ed episodes and I list all my folders with an anidb id

**Comment by ZeroQI** [comment]
@EndOfLine369 it only download if the anidb id is present and there are openings and wait 6s afterwards

**Comment by EndOfLine369** [comment]
ah, just rechecked. found the `time.sleep(6)` this time

**Comment by ZeroQI** [comment]
What is afterwards that test populate anidb-xml so it is downloaded only once in the file loop if needed
Your test is True if var was None and so was mine but you removed the warning and clearer, so better...

**Comment by EndOfLine369** [comment]
Looks like there are still issues in the test I have of openings.
https://anidb.net/perl-bin/animedb.pl?show=anime&aid=7506
There is a 'OP1d' in AniDB (that I don't have) that is not being reserved as e104.
```
AniDB URL: http://api.anidb.net:9001/httpapi?request=anime&client=hama&clientver=1&protover=1&aid=7506, length: 65620, AniDB_op: {1: {1: 3, 2: 4}}
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e151                         "ANIDB_RX-2" "Ending 1" "[Chihiro]_Amagami_SS_-_ED1_[1280x720_Blu-Ray_FLAC][970F7DCA].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e152                         "ANIDB_RX-2" "Ending 2" "[Chihiro]_Amagami_SS_-_ED2_[1280x720_Blu-Ray_FLAC][438325D1].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e153                         "ANIDB_RX-2" "Ending 3" "[Chihiro]_Amagami_SS_-_ED3_[1280x720_Blu-Ray_FLAC][6A1BF043].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e154                         "ANIDB_RX-2" "Ending 4" "[Chihiro]_Amagami_SS_-_ED4_[1280x720_Blu-Ray_FLAC][4F0C2A86].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e155                         "ANIDB_RX-2" "Ending 5" "[Chihiro]_Amagami_SS_-_ED5_[1280x720_Blu-Ray_FLAC][A71D8850].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e156                         "ANIDB_RX-2" "Ending 6" "[Chihiro]_Amagami_SS_-_ED6_[1280x720_Blu-Ray_FLAC][89893E4D].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e157                         "ANIDB_RX-2" "Ending 7" "[Chihiro]_Amagami_SS_-_ED7_[1280x720_Blu-Ray_FLAC][56FF6FAB].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e158                         "ANIDB_RX-2" "Ending 8" "[Chihiro]_Amagami_SS_-_ED8_[1280x720_Blu-Ray_FLAC][CD2A011D].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e101                         "ANIDB_RX-1" "Opening 1a" "[Chihiro]_Amagami_SS_-_OP1a_[1280x720_Blu-Ray_FLAC][90B8637C].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e102                         "ANIDB_RX-1" "Opening 1b" "[Chihiro]_Amagami_SS_-_OP1b_[1280x720_Blu-Ray_FLAC][1CB96475].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e103                         "ANIDB_RX-1" "Opening 1c" "[Chihiro]_Amagami_SS_-_OP1c_[1280x720_Blu-Ray_FLAC][AD82C54E].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e104                         "ANIDB_RX-1" "Opening 2a" "[Chihiro]_Amagami_SS_-_OP2a_[1280x720_Blu-Ray_FLAC][4B969037].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e105                         "ANIDB_RX-1" "Opening 2b" "[Chihiro]_Amagami_SS_-_OP2b_[1280x720_Blu-Ray_FLAC][DBDEB2E4].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e106                         "ANIDB_RX-1" "Opening 2c" "[Chihiro]_Amagami_SS_-_OP2c_[1280x720_Blu-Ray_FLAC][5CE1AD00].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e107                         "ANIDB_RX-1" "Opening 2d" "[Chihiro]_Amagami_SS_-_OP2d_[1280x720_Blu-Ray_FLAC][13BEC602].mkv"
"Amagami SS -lj -sU -e25 -a7506 [anidb-7506] -n11474" s00e108                         "ANIDB_RX-1" "Opening 2e" "[Chihiro]_Amagami_SS_-_OP2e_[1280x720_Blu-Ray_FLAC][6866FEAB].mkv"
```

**Comment by ZeroQI** [comment]
Look like it is in manual mode and doesn't use the mappings
First 1: is openings and 1:3 is ep 1 +3 (1d last op mapped) so cumulative offset should be 3 not 2 For op2x and the mapping were overwritten or the calculation array is wrong...

**Comment by EndOfLine369** [comment]
```
ep type offset: 100, ep: 1, offset: 0, cumulative_offset: 0, final ep number: 101
ep type offset: 100, ep: 1, offset: 1, cumulative_offset: 0, final ep number: 102
ep type offset: 100, ep: 1, offset: 2, cumulative_offset: 0, final ep number: 103
ep type offset: 100, ep: 2, offset: 0, cumulative_offset: 2, final ep number: 104
ep type offset: 100, ep: 2, offset: 1, cumulative_offset: 2, final ep number: 105
ep type offset: 100, ep: 2, offset: 2, cumulative_offset: 2, final ep number: 106
ep type offset: 100, ep: 2, offset: 3, cumulative_offset: 2, final ep number: 107
ep type offset: 100, ep: 2, offset: 4, cumulative_offset: 2, final ep number: 108
```

**Comment by EndOfLine369** [comment]
Looks like an issue in setting 'cumulative_offset' as should = 3 for ep2 openings

**Comment by EndOfLine369** [comment]
Think I have narrowed it down to this piece in issue.
```python
          if anidb_xml is not None:
            if ANIDB_RX.index(rx) in AniDB_op:  AniDB_op [ ANIDB_RX.index(rx) ]   [ ep ] = offset # {101: 0 for op1a / 152: for ed2b} and the distance between a and the version we have hereep, offset                         = str( int( ep[:-1] ) ), offset + sum( AniDB_op.values() )                             # "if xxx isdigit() else 1" implied since OP1a for example... # get the offset (100, 150, 200, 300, 400) + the sum of all the mini offset caused by letter version (1b, 2b, 3c = 4 mini offset)
            else:                               AniDB_op [ ANIDB_RX.index(rx) ] = { ep:    offset }
          Log.info("AniDB_op: {}".format(AniDB_op))
```
LOG:
```
AniDB URL: http://api.anidb.net:9001/httpapi?request=anime&client=hama&clientver=1&protover=1&aid=7506, length: 65620, AniDB_op: {1: {1: 3, 2: 4}}
AniDB_op: {1: {1: 3, 2: 4}, 2: {'1': 0}}
ep type offset: 150, ep: 1, offset: 0, cumulative_offset: 0, final ep number: 151
AniDB_op: {1: {1: 3, 2: 4}, 2: {'1': 0, '2': 0}}
ep type offset: 150, ep: 2, offset: 0, cumulative_offset: 0, final ep number: 152
AniDB_op: {1: {1: 3, 2: 4}, 2: {'1': 0, '3': 0, '2': 0}}
ep type offset: 150, ep: 3, offset: 0, cumulative_offset: 0, final ep number: 153
AniDB_op: {1: {1: 3, 2: 4}, 2: {'1': 0, '3': 0, '2': 0, '4': 0}}
ep type offset: 150, ep: 4, offset: 0, cumulative_offset: 0, final ep number: 154
AniDB_op: {1: {1: 3, 2: 4}, 2: {'1': 0, '3': 0, '2': 0, '5': 0, '4': 0}}
ep type offset: 150, ep: 5, offset: 0, cumulative_offset: 0, final ep number: 155
AniDB_op: {1: {1: 3, 2: 4}, 2: {'1': 0, '3': 0, '2': 0, '5': 0, '4': 0, '6': 0}}
ep type offset: 150, ep: 6, offset: 0, cumulative_offset: 0, final ep number: 156
AniDB_op: {1: {1: 3, 2: 4}, 2: {'1': 0, '3': 0, '2': 0, '5': 0, '4': 0, '7': 0, '6': 0}}
ep type offset: 150, ep: 7, offset: 0, cumulative_offset: 0, final ep number: 157
AniDB_op: {1: {1: 3, 2: 4}, 2: {'1': 0, '3': 0, '2': 0, '5': 0, '4': 0, '7': 0, '6': 0, '8': 0}}
ep type offset: 150, ep: 8, offset: 0, cumulative_offset: 0, final ep number: 158
AniDB_op: {1: {1: 0, 2: 4}, 2: {'1': 0, '3': 0, '2': 0, '5': 0, '4': 0, '7': 0, '6': 0, '8': 0}}
ep type offset: 100, ep: 1, offset: 0, cumulative_offset: 0, final ep number: 101
AniDB_op: {1: {1: 1, 2: 4}, 2: {'1': 0, '3': 0, '2': 0, '5': 0, '4': 0, '7': 0, '6': 0, '8': 0}}
ep type offset: 100, ep: 1, offset: 1, cumulative_offset: 0, final ep number: 102
AniDB_op: {1: {1: 2, 2: 4}, 2: {'1': 0, '3': 0, '2': 0, '5': 0, '4': 0, '7': 0, '6': 0, '8': 0}}
ep type offset: 100, ep: 1, offset: 2, cumulative_offset: 0, final ep number: 103
AniDB_op: {1: {1: 2, 2: 0}, 2: {'1': 0, '3': 0, '2': 0, '5': 0, '4': 0, '7': 0, '6': 0, '8': 0}}
ep type offset: 100, ep: 2, offset: 0, cumulative_offset: 2, final ep number: 104
AniDB_op: {1: {1: 2, 2: 1}, 2: {'1': 0, '3': 0, '2': 0, '5': 0, '4': 0, '7': 0, '6': 0, '8': 0}}
ep type offset: 100, ep: 2, offset: 1, cumulative_offset: 2, final ep number: 105
AniDB_op: {1: {1: 2, 2: 2}, 2: {'1': 0, '3': 0, '2': 0, '5': 0, '4': 0, '7': 0, '6': 0, '8': 0}}
ep type offset: 100, ep: 2, offset: 2, cumulative_offset: 2, final ep number: 106
AniDB_op: {1: {1: 2, 2: 3}, 2: {'1': 0, '3': 0, '2': 0, '5': 0, '4': 0, '7': 0, '6': 0, '8': 0}}
ep type offset: 100, ep: 2, offset: 3, cumulative_offset: 2, final ep number: 107
AniDB_op: {1: {1: 2, 2: 4}, 2: {'1': 0, '3': 0, '2': 0, '5': 0, '4': 0, '7': 0, '6': 0, '8': 0}}
ep type offset: 100, ep: 2, offset: 4, cumulative_offset: 2, final ep number: 108
```

**Comment by ZeroQI** [comment]
@EndOfLine369 
Two behaviours
- either there is an AniDB id and opening and the xml is loaded then pause 6s then build AniDB_op once for all
- OR it build manually from natural sorted titles. this cannot be applied to already build AniDB_op 
i then add the offset for all prior episodes (1a=0, ab=1, 1c=2...)

I believe you are better than me at python, but your pull request changed dramatically the behaviour: " - - Good: "not anidb_xml" replaced with "anidb_xml is None"
-Bad: "not anidb_xml"replaced with "anidb_xml is **not** None:". That is wrong. i didn't wrote it that way, and make the manual list over-write existing one and not update it for manual estimation when anidb id not present...

**Comment by EndOfLine369** [comment]
Episodes need at least a default offset entry of '0' otherwise a KeyError will be hit.
-------------------------
    cumulative_offset = sum( [ AniDB_op [ ANIDB_RX.index(rx) ][x] for x in Dict(AniDB_op, ANIDB_RX.index(rx), default={0:0}) if x<ep ] )
KeyError: (2,)
-------------------------
If a regex other than an op/en is matched, the 'ep_' variable is not known so it need to be updating the 'ep' variable instead of creating a new one.
-------------------------
    add_episode_into_plex(media, file, root, path, show, int(season), int(ep_), title, year, int(ep2) if ep2 and ep2.isdigit() else int(ep_), rx, tvdb_mapping, unknown_series_length, offset_season, offset_episode, mappingList)
UnboundLocalError: local variable 'ep_' referenced before assignment
-------------------------

**Comment by dgw** [comment]
@EndOfLine369 "fixes" from the top (summary) line combined with the issue number at the beginning of the description, which is unintuitive behavior at best.

Closing an issue referenced by number with "fixes", "closes", "resolves" etc. before it is documented GitHub behavior, but it probably shouldn't happen in this special case. I'd say it's worth writing to GitHub support.

**Comment by EndOfLine369** [comment]
Thanks @dgw for the explanation on the unexpected issue closure

**Comment by dgw** [comment]
Glad to help! I wrote GitHub support about it, not that I expect anything to change.

**Comment by EndOfLine369** [comment]
FYI, @ZeroQI,
Another crashe found
```python
    cumulative_offset = sum( [ AniDB_op [ ANIDB_RX.index(rx) ][x] for x in Dict(AniDB_op, ANIDB_RX.index(rx), default={0:0}) if x<ep ] )
KeyError: (0,)
```
Due to `if rx in ANIDB_RX[:-2]:` being  used but only `ANIDB_RX[1:3]` being used for `AniDB_op` population. Looks like the fix I put in earlier for the `KeyError: (2,)` crash needs a diff fix to handle any non-entry in `AniDB_op` for any of the `ANIDB_RX[:-2]` indexes.
Checking to see what is needed to fix.

**Comment by EndOfLine369** [comment]
Updated code to confirm 'AniDB_op's contents on the 'cumulative_offset=' line. seems to have fixed the issue.

**Comment by ZeroQI** [comment]
@purposelycryptic latest version is stable for you ?

**Comment by ZeroQI** [comment]
@purposelycryptic latest version is stable for you ?
Will close without reply
